File: frontend/src/utils/validators.ts
/**
 * Shared input validation utilities for frontend forms.
 *
 * These patterns mirror the backend validators in app/core/validators.py
 * to provide instant client-side feedback before API calls.
 */

// -- Regex patterns (matching backend SAFE_* patterns) --

/** Alphanumeric + dash + underscore + dot */
const SAFE_NAME = /^[a-zA-Z0-9_.\-]+$/;

/** IPv4/IPv6 address or CIDR */
const SAFE_IP = /^[a-fA-F0-9.:/%]+$/;

/** IP + optional hostname chars */
const SAFE_TARGET = /^[a-zA-Z0-9._:\-]+$/;

/** BPF filter safe chars */
const SAFE_BPF = /^[a-zA-Z0-9 ._:\-/()!=<>&|]+$/;

/** Simple IPv4 format (not exhaustive, just basic shape check) */
const IPV4_PATTERN = /^(\d{1,3}\.){3}\d{1,3}(\/\d{1,2})?$/;

// -- Validator functions --

export interface ValidationResult {
  valid: boolean;
  error?: string;
}

const ok: ValidationResult = { valid: true };
const err = (msg: string): ValidationResult => ({ valid: false, error: msg });

/**
 * Validate an identifier name (switch, VRF, interface, etc.)
 */
export function validateName(value: string, field = 'Name'): ValidationResult {
  const v = value.trim();
  if (!v) return err(`${field} is required`);
  if (v.length > 64) return err(`${field} is too long (max 64 characters)`);
  if (!SAFE_NAME.test(v)) return err(`${field} may only contain letters, numbers, dash, underscore, and dot`);
  return ok;
}

/**
 * Validate an IP address or CIDR notation.
 */
export function validateIP(value: string, field = 'IP address'): ValidationResult {
  const v = value.trim();
  if (!v) return err(`${field} is required`);
  if (v.length > 45) return err(`${field} is too long`);
  if (!SAFE_IP.test(v)) return err(`${field} contains invalid characters`);
  return ok;
}

/**
 * Validate an IPv4 address with optional CIDR prefix.
 */
export function validateIPv4(value: string, field = 'IP address'): ValidationResult {
  const v = value.trim();
  if (!v) return err(`${field} is required`);
  if (!IPV4_PATTERN.test(v)) return err(`${field} must be a valid IPv4 address (e.g. 10.0.0.1 or 10.0.0.0/24)`);

  // Check each octet is 0-255
  const [ip, prefix] = v.split('/');
  const octets = ip.split('.').map(Number);
  if (octets.some((o) => o < 0 || o > 255)) return err(`${field} has invalid octets (must be 0-255)`);
  if (prefix !== undefined) {
    const p = Number(prefix);
    if (p < 0 || p > 32) return err(`${field} has invalid prefix length (must be 0-32)`);
  }
  return ok;
}

/**
 * Validate a ping/traceroute target (IP or hostname).
 */
export function validateTarget(value: string, field = 'Target'): ValidationResult {
  const v = value.trim();
  if (!v) return err(`${field} is required`);
  if (v.length > 253) return err(`${field} is too long`);
  if (!SAFE_TARGET.test(v)) return err(`${field} contains invalid characters`);
  return ok;
}

/**
 * Validate an ASN (Autonomous System Number).
 */
export function validateASN(value: number | string, field = 'ASN'): ValidationResult {
  const n = typeof value === 'string' ? parseInt(value, 10) : value;
  if (isNaN(n)) return err(`${field} must be a number`);
  if (n < 1 || n > 4294967295) return err(`${field} must be between 1 and 4,294,967,295`);
  return ok;
}

/**
 * Validate a description field.
 */
export function validateDescription(value: string, field = 'Description'): ValidationResult {
  if (!value) return ok; // descriptions are typically optional
  if (value.length > 200) return err(`${field} is too long (max 200 characters)`);
  return ok;
}

/**
 * Validate a BPF filter expression.
 */
export function validateBPF(value: string, field = 'BPF filter'): ValidationResult {
  if (!value) return ok;
  if (value.length > 500) return err(`${field} is too long`);
  if (!SAFE_BPF.test(value)) return err(`${field} contains invalid characters`);
  return ok;
}

File: backend/app/core/validators.py
"""Shared input validation utilities to prevent command injection.

All user-supplied values that will be interpolated into SSH/shell commands
MUST be validated through these helpers before use.
"""

from __future__ import annotations

import re

from fastapi import HTTPException

# -- Compiled patterns --------------------------------------------------------

# Alphanumeric + dash + underscore + dot (names, identifiers, VLAN ifaces)
SAFE_NAME = re.compile(r"^[a-zA-Z0-9_.\-]+$")

# IPv4/IPv6 address or CIDR (e.g. 10.0.0.1, 10.0.0.0/24, ::1)
SAFE_IP = re.compile(r"^[a-fA-F0-9.:/%]+$")

# IP + optional hostname chars (for targets like hostnames)
SAFE_TARGET = re.compile(r"^[a-zA-Z0-9._:\-]+$")

# BPF filter expressions (tcpdump)
SAFE_BPF = re.compile(r"^[a-zA-Z0-9 ._:\-/()!=<>&|]+$")

# ASN: positive integer
SAFE_ASN = re.compile(r"^\d+$")

# VRF/router-id style: dotted quad or simple name
SAFE_ROUTER_ID = re.compile(r"^[a-fA-F0-9.:]+$")

# Description: printable ASCII, no shell metacharacters
SAFE_DESCRIPTION = re.compile(r'^[a-zA-Z0-9 _\-.,/()#@]+$')


# -- Validator functions ------------------------------------------------------

def validate_name(value: str, field: str = "name") -> str:
    """Validate an identifier name (alphanumeric, dash, underscore)."""
    value = value.strip()
    if not value or not SAFE_NAME.match(value):
        raise HTTPException(400, detail=f"Invalid {field}: {value!r} — only alphanumeric, dash, underscore, dot allowed")
    if len(value) > 64:
        raise HTTPException(400, detail=f"{field} too long (max 64 chars)")
    return value


def validate_ip(value: str, field: str = "IP address") -> str:
    """Validate an IP address or CIDR notation."""
    value = value.strip()
    if not value or not SAFE_IP.match(value):
        raise HTTPException(400, detail=f"Invalid {field}: {value!r}")
    if len(value) > 50:
        raise HTTPException(400, detail=f"{field} too long")
    return value


def validate_target(value: str, field: str = "target") -> str:
    """Validate a target (IP address or hostname)."""
    value = value.strip()
    if not value or not SAFE_TARGET.match(value):
        raise HTTPException(400, detail=f"Invalid {field}: {value!r}")
    if len(value) > 253:
        raise HTTPException(400, detail=f"{field} too long")
    return value


def validate_asn(value: int, field: str = "ASN") -> int:
    """Validate an Autonomous System Number."""
    if not (1 <= value <= 4294967295):
        raise HTTPException(400, detail=f"Invalid {field}: {value} — must be 1..4294967295")
    return value


def validate_description(value: str, field: str = "description") -> str:
    """Validate a description string (no shell metacharacters)."""
    value = value.strip()
    if not value:
        return value
    if not SAFE_DESCRIPTION.match(value):
        raise HTTPException(400, detail=f"Invalid {field}: contains disallowed characters")
    if len(value) > 200:
        raise HTTPException(400, detail=f"{field} too long (max 200 chars)")
    return value
